src/SAL/recovery/recovery_longhorn.c

```diff
@@ -0,0 +1,721 @@
+	if (len <= 0 || len > max)
+		return 0;
+
+	/* If the value is NULL, display NULL value. */
+	if (value == NULL)
+		return 0;
+
+	/* Determine if the value is entirely printable characters, */
+	/* and it contains no slash character (reserved for filename) */
+	for (i = 0; i < len; i++)
+		if ((!isprint(((char *)value)[i])) ||
+		    (((char *)value)[i] == '/'))
+			break;
+
+	if (i == len) {
+		/* Entirely printable character, so we will just copy the
+		 * characters into the buffer (to the extent there is room
+		 * for them).
+		 */
+		b_left = display_len_cat(dspbuf, value, cpy);
+	} else {
+		b_left = display_opaque_bytes(dspbuf, value, cpy);
+	}
+
+	if (b_left <= 0)
+		return 0;
+
+	return b_left;
+}
+
+/**
+ * @brief generate a name that identifies this client
+ *
+ * This name will be used to know that a client was talking to the
+ * server before a restart so that it will be allowed to do reclaims
+ * during grace period.
+ *
+ * @param[in] clientid Client record
+ */
+static void longhorn_create_clid_name(nfs_client_id_t *clientid)
+{
+	nfs_client_record_t *cl_rec = clientid->cid_client_record;
+	const char *str_client_addr = "(unknown)";
+	char cidstr[PATH_MAX] = { 0, };
+	struct display_buffer dspbuf = {sizeof(cidstr), cidstr, cidstr};
+	char cidstr_lenx[5];
+	int total_size, cidstr_lenx_len, cidstr_len, str_client_addr_len;
+
+	/* get the caller's IP addr */
+	if (clientid->gsh_client != NULL)
+		str_client_addr = clientid->gsh_client->hostaddr_str;
+
+	if (longhorn_convert_opaque_value_max_for_dir(&dspbuf,
+						cl_rec->cr_client_val,
+						cl_rec->cr_client_val_len,
+						PATH_MAX) > 0) {
+		cidstr_len = strlen(cidstr);
+		str_client_addr_len = strlen(str_client_addr);
+
+		/* longhorn_convert_opaque_value_max_for_dir does not prefix
+		 * the "(<length>:". So we need to do it here */
+		cidstr_lenx_len = snprintf(cidstr_lenx, sizeof(cidstr_lenx),
+					   "%d", cidstr_len);
+
+		if (unlikely(cidstr_lenx_len >= sizeof(cidstr_lenx) ||
+			     cidstr_lenx_len < 0)) {
+			/* cidrstr can at most be PATH_MAX or 1024, so at most
+			 * 4 characters plus NUL are necessary, so we won't
+			 * overrun, nor can we get a -1 with EOVERFLOW or EINVAL
+			 */
+			LogFatal(COMPONENT_CLIENTID,
+				 "snprintf returned unexpected %d",
+				 cidstr_lenx_len);
+		}
+
+		total_size = cidstr_len + str_client_addr_len + 5 +
+			     cidstr_lenx_len;
+
+		/* hold both long form clientid and IP */
+		clientid->cid_recov_tag = gsh_malloc(total_size);
+
+		/* Can't overrun and shouldn't return EOVERFLOW or EINVAL */
+		(void) snprintf(clientid->cid_recov_tag, total_size,
+				"%s:%s",
+				cidstr_lenx, cidstr);
+	}
+
+	LogDebug(COMPONENT_CLIENTID, "Created client name [%s]",
+		 clientid->cid_recov_tag);
+}
+
+static int longhorn_recov_init(void)
+{
+	char host[NI_MAXHOST];
+	char payload[PAYLOAD_MAX];
+	char *response = NULL;
+	size_t response_size = 0;
+	char *version = NULL;
+	int err = 0;
+	int res = 0;
+
+	err = gethostname(host, sizeof(host));
```

Comment:
Will `host` keep the same after the share manager pod failed over?

Reply:
Yes,
The hostname inside the share-manager pod is same as the name of the share-manager pod.
It is not changed after failover.

Reply:
ref: https://github.com/longhorn/longhorn-manager/blob/c3e9b1c8c7d1043134bfbf2c8ceb25c789935bc0/controller/share_manager_controller.go#L784-L784